File: Backend/src/modules/customers/models/customer.model.ts
import { Schema, model, type Document, type Model } from 'mongoose';

export interface ICustomer {
  fullName: string;
  phone: string;
  address: string;
}

export interface ICustomerDocument extends ICustomer, Document {
  createdAt: Date;
  updatedAt: Date;
}

const customerSchema = new Schema<ICustomerDocument>(
  {
    fullName: {
      type: String,
      required: true,
      trim: true,
      minlength: 2,
      maxlength: 120,
      index: true,
    },
    phone: {
      type: String,
      required: true,
      trim: true,
      maxlength: 32,
      index: true,
    },
    address: {
      type: String,
      required: true,
      trim: true,
      maxlength: 240,
    },
  },
  {
    timestamps: true,
    versionKey: false,
  },
);

customerSchema.index({ fullName: 'text', phone: 'text', address: 'text' });

export const CustomerModel: Model<ICustomerDocument> = model<ICustomerDocument>('Customer', customerSchema);

File: lib/features/customers/data/models/customer_model.dart
final class CustomerModel {
  const CustomerModel({
    required this.id,
    required this.fullName,
    required this.phone,
    required this.address,
    required this.createdAt,
    required this.updatedAt,
  });

  final String id;
  final String fullName;
  final String phone;
  final String address;
  final DateTime createdAt;
  final DateTime updatedAt;

  factory CustomerModel.fromJson(Map<String, dynamic> json) {
    return CustomerModel(
      id: json['id'] as String,
      fullName: json['fullName'] as String,
      phone: json['phone'] as String,
      address: json['address'] as String,
      createdAt: DateTime.parse(json['createdAt'] as String),
      updatedAt: DateTime.parse(json['updatedAt'] as String),
    );
  }
}
